### ISSUE TITLE
Communicating client tag deny/allow lists to clients

### ISSUE THREAD

**Issue by jwheare** [ircv3/ircv3-specifications#410]
Following on from https://github.com/ircv3/ircv3-specifications/issues/333#issuecomment-491496090

Since unreal (@syzop) and now inspircd (@SadieCat - https://github.com/inspircd/inspircd/commit/7488b2de62811d7f831e84fb0aeac2345a7f9ee9) both allow or enforce client tag disabling, we need a solution for clients that need to know which client tags can be used, so they can show e.g. a reply ui to users.

The use cases are, in order of how likely they are:

1. All allowed (default)
2. All blocked (catch-all deny)
3. Some allowed (catch-all deny + allow)
4. Some blocked (deny)

A few decisions:

* Where should this information be published? In ISUPPORT or the tag value or via a negotiated command?
* What format should be used?
    * What to call the key?
    * Do we need to include the + when listing tags?
    * Can we use * for a catch-all?
    * Should we allow * as a wildcard, e.g. foo*?
    * Can we use ! for negation?
    * Can we use !* for both?
    * Can we use , or ; for separating rules?
    * Should allow and deny rules be listed in separate keys?
* Which order are deny/allow/defaults applied in?
* How are empty or omitted deny/allow rules interpreted?

Here are some examples of how this might look, using isupport. If no key is present, the default behaviour is always 1) All allowed.

# Single allow key `CLIENTTAGALLOW`
If omitted *or empty* defaults to *, all allowed

1. `CLIENTTAGALLOW=*`
2. `CLIENTTAGALLOW=!*` (negative catch-all)
3. `CLIENTTAGALLOW=!*,foo` (negative catch-all)
4. `CLIENTTAGALLOW=!foo`

# Single deny key `CLIENTTAGDENY`
If omitted, defaults to empty, nothing blocked

1. `CLIENTTAGDENY=`
2. `CLIENTTAGDENY=*`
3. `CLIENTTAGDENY=*,!foo` (double-negative)
4. `CLIENTTAGDENY=foo`

# Separate keys `CLIENTTAGALLOW` `CLIENTTAGDENY`
Same default rules apply as above: `CLIENTTAGALLOW` defaults to *, including when empty and `CLIENTTAGDENY` defaults to empty

Allow applied first:

1. `CLIENTTAGALLOW=* CLIENTTAGDENY=`
2. `CLIENTTAGALLOW= CLIENTTAGDENY=*`
3. `CLIENTTAGALLOW=foo CLIENTTAGDENY=!*` (negative catch-all, double-negative)
4. `CLIENTTAGALLOW=* CLIENTTAGDENY=foo`

Deny applied first:

1. `CLIENTTAGDENY= CLIENTTAGALLOW=*`
2. `CLIENTTAGDENY=* CLIENTTAGALLOW=!*` (negative catch-all)
3. `CLIENTTAGDENY=* CLIENTTAGALLOW=foo`
4. `CLIENTTAGDENY=foo CLIENTTAGALLOW=!*` (negative catch-all)

---

I think the format of a single `CLIENTTAGDENY` makes the most sense, doesn't rely on a negative catch-all and doesn't present rule interpretation order issues.

(As long as we assume `CLIENTTAGDENY=*,!foo` is the same as `CLIENTTAGDENY=!foo,*`, i.e. negated terms are always interpreted last irrespective of list ordering)

So looking again at the list of questions/decisions, some remain undecided but assuming we use the format/protocol/naming from these example, this is my opinion:

* Where: ISUPPORT (undecided)
* What format should be used:
    * What to call the key: `CLIENTTAGDENY` (undecided)
    * Do we need to include the + when listing tags (no)
    * Can we use * for a catch-all (yes)
    * Should we allow * as a wildcard, e.g. foo* (no)
    * Can we use ! for negation (yes)
    * Can we use !* for both (no)
    * Can we use , or ; for separating rules (yes, comma is fine)
    * Should allow and deny rules be listed in separate keys (no)
* Which order are deny/allow/defaults applied in (default -> deny -> negated-allow)
* How are empty or omitted deny/allow rules interpreted (default, all allowed, nothing blocked)

Comments, agree/disagreements, thoughts on the undecided parts welcome, *especially* from people who've already gone ahead and started allowing people to deny tags in their ircds...

**Comment by SadieCat** [comment]
It'd be simpler to just have one token and allow negation, like for a blacklist:

```
CLIENTTAGS=*,-foo,-bar-baz
```

And for a whitelist:

```
CLIENTTAGS=-*,foo,bar,baz
```

**Comment by jwheare** [comment]
That's the same as `CLIENTTAGALLOW` with - instead of !

I don't like the negative catch-all (-*) that's needed for cases 2 (block all) and 3 (allow some)

**Comment by prawnsalad** [comment]
A spec to cripple another spec.

While disabling client tags entirely I can (just about) understand and is clear-cut - the network doesn't want any hidden bits of info being sent between clients, the idea of this spec will only confuse and over complicate things on the client side.

The whole idea of client tags is that clients have more freedom to experiment and trial out new functionality rather than waiting for servers to notice, acknowledge, and then configure/implement the feature. Either disable them entirely so that clients know to only use bare minimum functionality and can warn the user, or allow them as the message-tags spec defines so that users have a choice themselves. This whole murky middle-ground would force a client to check what tags are allowed then negotiate within itself which features can still work, and then warn the user that feature x/y/z cannot be used. That's a terrible experience.

Why is it that network admins stuck in the telnet days are forcing their mindset onto all of their users? Let the users decide what they want - it's their client. If I as a user decide that I'm getting abused by a tag then I can either report it to network/channel staff and the useable abuse routine follows, or I disable tags myself. I don't want someone else to tell me what's good for me.

**Comment by syzop** [comment]
@prawnsalad I think the whole reason jwheare spun this issue off from https://github.com/ircv3/ircv3-specifications/issues/333 is to avoid going through that discussion again.

I agree with @SadieCat to have just one token. You would then have to decide to default to whitelist or blacklist, or at least if you want to keep things simple. I suggest to default to whitelist because then you can use *, while with blacklisting it becomes harder and you get double-negation and stuff.

So, very similar to what @SadieCat said (and incorporating basically what @jwheare said):
Allow all except 'foo', 'bar' and 'baz':
```CLIENTTAGS=*,-foo,-bar,-baz```
And allow only 'foo', 'bar', and 'baz':
```CLIENTTAGS=foo,bar,baz```

**Comment by syzop** [comment]
Then, as to whether it should be a CAP or an ISUPPORT token:
I suppose the general question for any feature is: does the client needs to know about this at that early connect stage. The only thing I can think of is if you need to know about client-tag support in the handshake, that is SASL or some other handshake stuff to a server or services. That sounds a bit far-fetched to me, but that is the only reason I could possibly think of. For all the rest you can wait until after the 005 ISUPPORT.
So, I would go with ISUPPORT unless you think there is some possible scenario where you could possibly need it at the handshake stage.
On a side note, as a server-developer I don't care much about where it ends up with, ISUPPORT or CAP, so I am fine with whatever preference from client devs or anyone else.

**Comment by prawnsalad** [comment]
> @prawnsalad I think the whole reason jwheare spun this issue off from #333 is to avoid going through that discussion again.

You may be in favour of forcing client functionality upon users but as someone that would have to implement this functionality and *knowing* it will confuse IRC users (and many already have been impacted by Unreals heavy handed approach to limiting users choice), I don't think the original conversation those years ago even touched the surface.

Nowhere did it mention these drawbacks, nowhere was the "security" aspect actually described.

Note that I'm not saying client-tags should be enabled at all time. I'm saying that this middle ground of a mix of allowed client tags will only cause confusion and terrible experience for users. Pick one or the other.

**Comment by realJoshByrnes** [comment]
I definitely agree with @prawnsalad on this one.

The two issues I've heard of go something like this:
* User X is sending hidden messages to the channel using client tags
Disabling one tag won't stop them from repurposing another tag (eg.; Typing indicators). If you're really concerned, disable all client tags.
* There's too much network activity
Again, filtering some client tags won't fix this. Just disable all client tags.

It might seem trivial, but in reality, it's going to become a burden on all developers (server/client/user) as well as being almost pointless to try and work around the above issues. (kinda like trying to stop someone from writing something mean in PRIVMSG)

The only other way to sort this out is to only allow certain values (eg. typing indicators = typing / not typing), which would create further overheads as well as pretty much break the whole point of having client tags.

**Comment by DanielOaks** [comment]
I reckon the syntax @SadieCat suggests looks pretty sane. We won't ship with any restrictions but if it helps other servers adopt client tags then awesome.

**Comment by SadieCat** [comment]
I should clarify, I was just giving a suggestion earlier not endorsing nor opposing any particular choice.

We have no plans to implement fine grained tag restrictions so i'd be fine with a boolean on/off field if thats what works for client developers.

**Comment by jwheare** [comment]
Another reason I much prefer the token to be deny rather than allow is that the default case (empty, omitted) matches the actual default which is to allow all. I want to reinforce very strongly that you're doing something out of the ordinary by filtering client tags, not advertise a token that looks like it's mandatory for the feature.

I feel like it's simpler to deal with deny rules being the exception from an implementation perspective too.

**Comment by jwheare** [comment]
@prawnsalad:
> Either disable them entirely so that clients know to only use bare minimum functionality and can warn the user, or allow them as the message-tags spec defines so that users have a choice themselves. This whole murky middle-ground would force a client to check what tags are allowed then negotiate within itself which features can still work, and then warn the user that feature x/y/z cannot be used. That's a terrible experience.

I don't see any difference in confusion between all tags being disabled and only some. You still have to check and deal with the feature being disabled somehow.

Anyway, this isn't really about whether we like or dislike this status quo, it's about giving clients a way to deal with the situation that's already happening in the wild. There are servers that run on unreal's tags-without-client-tags but have modules that selectively allow them, e.g. +typing tags. I can quite easily imagine more tags being given this treatment in future.

**Comment by jwheare** [comment]
Small thing, but just a note that I agree with suggestions that - is better than ! for negation.

**Comment by syzop** [comment]
I think if you default to blacklist, then the name of the token/cap should also reflect that.
So, yeah, in terms of consensus I would then go with ```CLIENTTAGDENY``` what @jwheare proposes.
Then you still have the question of ```!``` or ```-``` for negating. I really couldn't care less, so just pick one.
As for CAP vs ISUPPORT... I can't think of a reason to go with CAP indeed, so I would say ISUPPORT unless someone posts here a good reason to choose CAP.

**Comment by jwheare** [comment]
The only reason to use cap values would be if this is information needed during the client-server opening handshake. Since this is about client-client tags only, I'm pretty sure that will never be the case. (Note: non-client tags might still be usable during c2s handshake)

**Comment by slingamn** [comment]
1. As an operator, I respect the desire of operators to be able to disable client-only tags.
2. As a server vendor, I'm sympathetic enough to this desire that I would consider adding it as an optional feature.
3. As a specs person, and as someone concerned about the overall ecosystem, I think the behavior of such a server is clearly non-compliant with the existing [message-tags](https://ircv3.net/specs/extensions/message-tags.html) specification: "Client-only tags are tags normally sent by clients that servers MUST attach as-is to any relevant message relayed to other clients." Moreover, I think this configuration is too marginal to justify an erratum to the existing specification.
4. Therefore, I feel that the way to advertise this behavior to clients is simply not to advertise the `message-tags` cap. So this is a vote for a binary choice: either relay all client-only tags (and advertise `message-tags`), or none of them (and don't).
5. It should still be possible for these configurations to advertise caps that use (server) tags. In particular, `server-time` and `account-tag` predated the ratified tags specification and are expected to work without negotiating `message-tags`. With any future capability that depends on server-side tags, we could put the following sort of language in its specification: "tags MUST be parsed according to the ratified message-tags specification, but this capability does not depend on message-tags being negotiated, i.e., if the client has negotiated this capability but not message-tags, the capability MUST be enabled regardless."

**Comment by slingamn** [comment]
After some private discussion, I'm not really set on the "binary choice" aspect of this --- I don't really have anything against servers implementing whitelists and/or blacklists. But I still think that if any such restrictions are imposed by the server, `message-tags` shouldn't be advertised.

**Comment by SadieCat** [comment]
> But I still think that if any such restrictions are imposed by the server, message-tags shouldn't be advertised.

This removes the ability to use msgid, labeled-response, and every other non-client only tag spec gated on havng the message-tags caps. No thanks.

**Comment by jwheare** [comment]
There are server tags that get sent when message-tags cap is enabled that don’t have their own cap. Eg msgid.

**Comment by slingamn** [comment]
I withdraw my previous comments.

My main concern here is that this erratum shouldn't impose any additional burden on implementers (in particular, people shouldn't feel like they have to parse the tokens). It sounds like this will be the case.

**Comment by jwheare** [comment]
Yup that is the case.

Appreciate the discussion around this. I’ll PR a version that I think has general consensus.

**Comment by jwheare** [comment]
PR opened at #412
